Standards-Code-Formatting.md

```diff
@@ -0,0 +1,377 @@
+The key here is that writing readable code always takes priority.
+
+### Vertical Whitespace
+
+Always place a single blank line after a control structure unless the following
+line is an ending brace `}` or is the last line in the file.
+
+```php
+if ($var == 0) {
+    $i = 0;
+}
+
+if ($i == 0) {
+    $i *= 2;
+}
+```
+
+```php
+if ($var ==0) {
+    if ($i == 2) {
+        $i++;
+    }
+}
+```
+
+## Storage Constructs
+
+### Constants
+
+Always use constants in place of numbers to represent static values. Instead of
+doing:
+
+```php
+if ($poller_verbosity == 2) {
+    // code
+}
+```
+
+Do:
+
+```php
+if ($poller_verbosity == POLLER_VERBOSITY_LOW) {
+    // code
+}
+```
+
+### Variables
+
+Allocating memory for a variable consumes resources. Only declare a variable if
+it is referenced more than once throughout the code. Instead of doing:
+
+```php
+$num_devices = device_total_num_get();
+
+if ($num_devices > 0) {
+    // code
+}
+```
+
+Do:
+
+```php
+if (device_total_num_get() > 0) {
+    // code
+}
+```
+
+### Global Variables
+
+Certain situations warrant creating a single variable that is to be referenced
+by multiple functions. Do not rely on PHP's `global` keyword since it creates
+the possibility of name space collisions. Instead, create a wrapper function
+that brings the variable into a global scope and returns a new copy to any
+function that needs to use it.
+
+```php
+$graph_unit_exponent_values = array('-18' => 'a - atto', '-15' => 'f - femto', ...);
+
+function graph_unit_exponent_values_list() {
+    global $graph_unit_exponent_values;
+
+    return $graph_unit_exponent_values;
+}
+
+function graph_unit_exponent_values_print() {
+    foreach (graph_unit_exponent_values_list() as $power => $unit) {
+        // code
+    }
+}
+```
+
+## References
+
+Copying around large arrays in PHP can be an expensive operation. Be sure to
+return or pass arrays by reference when data corruption is not a huge concern.
+Keep in mind that call time pass by reference is deprecated in PHP.
+
+```php
+function &big_array_list() {
+    // code
+}
+```
+
+```php
+function big_array_process(&$big_array) {
+    // code
+}
+```
+
+## Comments
+
+The key to being a respectable coder is to write clear, concise code and to use
+comments everywhere.
+
+### Whole Line Comments
+
+All single and multi-line comments should follow the C-style syntax.
+
+```php
+/* this determines if $i is equal to 0 */
+if ($i == 0) {
+    // code
+}
+```
+
+```php
+/* this is an extra long comment that deserves to be
+ * split up into multiple lines */
+if ($i == 0) {
+    // code
+}
+```
+
+### End of Line Comments
+
+Shorter comments that follow a line of code should use the familiar `//` style.
+
+```php
+$var = 0; // assign 0 to $var
+```
+
+### File Header Comments
+
+All code related files must contain the following notice at the beginning of
+the file.
+
+```php
+/*
+ +-------------------------------------------------------------------------+
```

Comment:
This should be updated with the latest copyright notice.  We can do this after the merge if there are no other issues.